sapi/fuzzing/fuzzer-sapi.c

```diff
@@ -0,0 +1,257 @@
+};
+
+int fuzzer_init_php()
+{
+	sapi_startup(&fuzzer_module);
+	fuzzer_module.phpinfo_as_text = 1;
+
+	fuzzer_module.ini_entries = malloc(sizeof(HARDCODED_INI));
+	memcpy(fuzzer_module.ini_entries, HARDCODED_INI, sizeof(HARDCODED_INI));
+
+	/*
+	 * TODO: we might want to test both Zend and malloc MM, but testing with malloc
+	 * is more likely to find bugs, so use that for now.
+	 */
+	putenv("USE_ZEND_ALLOC=0");
+
+#ifdef __SANITIZE_ADDRESS__
+	/* Not very interested in memory leak detection, since Zend MM does that */
+	__lsan_disable();
+#endif
+
+	if (fuzzer_module.startup(&fuzzer_module)==FAILURE) {
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+void fuzzer_set_ini_file(const char *file)
+{
+	if (fuzzer_module.php_ini_path_override) {
+		free(fuzzer_module.php_ini_path_override);
+	}
+	fuzzer_module.php_ini_path_override = strdup(file);
+}
+
+
+int fuzzer_shutdown_php()
+{
+	TSRMLS_FETCH();
+
+	php_module_shutdown(TSRMLS_C);
+	sapi_shutdown();
+
+	free(fuzzer_module.ini_entries);
+	return SUCCESS;
+}
+
+int fuzzer_do_request(zend_file_handle *file_handle, char *filename)
+{
+	int retval = FAILURE; /* failure by default */
+
+	SG(options) |= SAPI_OPTION_NO_CHDIR;
+	SG(request_info).argc=0;
+	SG(request_info).argv=NULL;
+
+	if (php_request_startup(TSRMLS_C)==FAILURE) {
+		php_module_shutdown(TSRMLS_C);
+		return FAILURE;
+	}
+
+	SG(headers_sent) = 1;
+	SG(request_info).no_headers = 1;
+	php_register_variable("PHP_SELF", filename, NULL TSRMLS_CC);
+
+	zend_first_try {
+		zend_compile_file(file_handle, ZEND_REQUIRE);
+		/*retval = php_execute_script(file_handle TSRMLS_CC);*/
+	} zend_end_try();
+
+	php_request_shutdown((void *) 0);
+
+	return (retval == SUCCESS) ? SUCCESS : FAILURE;
+}
+
+
+int fuzzer_do_request_f(char *filename)
+{
+	zend_file_handle file_handle;
+	file_handle.type = ZEND_HANDLE_FILENAME;
+	file_handle.filename = filename;
+	file_handle.handle.fp = NULL;
+	file_handle.opened_path = NULL;
+
+	return fuzzer_do_request(&file_handle, filename);
+}
+
+int fuzzer_do_request_d(char *filename, char *data, size_t data_len)
+{
+	zend_file_handle file_handle;
+	file_handle.filename = filename;
+	file_handle.opened_path = NULL;
+	file_handle.handle.stream.handle = NULL;
+	file_handle.handle.stream.reader = (zend_stream_reader_t)_php_stream_read;
+	file_handle.handle.stream.fsizer = NULL;
+	file_handle.handle.stream.isatty = 0;
+	file_handle.handle.stream.closer   = NULL;
+	file_handle.buf = data;
+	file_handle.len = data_len;
+	file_handle.type = ZEND_HANDLE_STREAM;
+
+	return fuzzer_do_request(&file_handle, filename);
+}
+
+// Call named PHP function with N zval arguments
+void fuzzer_call_php_func_zval(const char *func_name, int nargs, zval *args) {
+	zval retval, func;
+	int result;
+
+	ZVAL_STRING(&func, func_name);
+	ZVAL_UNDEF(&retval);
+	result = call_user_function(CG(function_table), NULL, &func, &retval, nargs, args);
+
+	// TODO: check result?
+	/* to ensure retval is not broken */
+	php_var_dump(&retval, 0);
```

Comment:
I'm not sure I understand the purpose of this var_dump call.

Reply:
This is to ensure the variable is not messed up. If it is, var_dump hopefully would stumble on it.